### PR TITLE
Feature/WOM-115: Featured Stories Carousel - Administrator

### PR DESCRIPTION
**Pull request by TraceyKong**
Added carousel to home page, edited feature page so ranks can be editted

### CODE REVIEW

File: app/feature/utils.py [hunk 1]
```diff
@@ -8,27 +8,32 @@
...
 
 
-def create_featured_story(story, left_right, quote):
+def create_featured_story(story, left_right, title, description, rank):
     """
     A utility function to create a featured story.
     :param story: the story object you would like to create in Featured Story
     :param left_right: left or right side where the image will go
-    :param quote: a famous quote that was once said by this activist
+    :param description: description of the activist
```
Comment (zgary) [inline app/feature/utils.py]: Add a docstring for `title`.

File: app/feature/utils.py [hunk 2]
```diff
@@ -8,27 +8,32 @@
...
+    :param rank: the order of the featured story
+
     :return: None
     """
-    #
-    current_featured_story = FeaturedStories.query.filter_by(is_visible=True).one_or_none()
-    if current_featured_story is not None:
-        update_object({"is_visible": False}, FeaturedStories, current_featured_story.id)
+
+    #current_featured_story = FeaturedStories.query.filter_by(is_visible=True)
```
Comment (zgary) [inline app/feature/utils.py]: Remove if not being used.

File: app/feature/utils.py [hunk 3]
```diff
@@ -8,27 +8,32 @@
...
         left_right=True if left_right == 'left' else False,
         is_visible=True,
-        quote=quote,
+        title=title,
+        description=description,
+        rank=rank
     )
 
     create_object(featured_story)
+    update_rank(featured_story.story_id, None, rank)
```
Comment (zgary) [inline app/feature/utils.py]: Why is this needed upon creating the featured story?

File: app/feature/utils.py [hunk 4]
```diff
@@ -38,27 +43,32 @@ def create_featured_story(story, left_right, quote):
...
-def update_featured_story(featured_story, left_right, is_visible, quote):
+def update_featured_story(featured_story, left_right, title, is_visible, description, rank):
     """
     A utility function to update a featured story.
     Updating attributes such as the following:
     :param featured_story:
     :param left_right: left or right side where the image will go
     :param is_visible: the visibility of the featured story
-    :param quote: a famous quote that was once said by this activist
+    :param description: description of the activist
```
Comment (zgary) [inline app/feature/utils.py]: Add docstring for `title`.

File: app/feature/utils.py [hunk 5]
```diff
@@ -74,13 +84,16 @@ def update_featured_story(featured_story, left_right, is_visible, quote):
             if cur_val != new_val:
                 old[field] = cur_val
                 new[field] = new_val
+                print(new_val)
```
Comment (zgary) [inline app/feature/utils.py]: Remove if only used for testing purposes.

File: app/feature/utils.py [hunk 6]
```diff
@@ -74,13 +84,16 @@ def update_featured_story(featured_story, left_right, is_visible, quote):
...
 
     if new:
-        if new.get('is_visible'):
-            hide_current_featured_story()
+        cur_rank = getattr(featured_story, "rank")
```
Comment (zgary) [inline app/feature/utils.py]: You can use `featured_story.rank` to get the current rank.

File: app/feature/utils.py [hunk 7]
```diff
@@ -90,21 +103,61 @@ def update_featured_story(featured_story, left_right, is_visible, quote):
...
+    :param: story_id: story id of the featured story that will be hidden
 
     :return: None
     """
-    current_featured_story = FeaturedStories.query.filter_by(is_visible=True).one_or_none()
-    if current_featured_story is not None:
-        update_object({"is_visible": False}, FeaturedStories, current_featured_story.id)
+    featured_story = FeaturedStories.query.filter_by(story_id=story_id, is_visible=True).one_or_none()
+    
+    # current_featured_story = FeaturedStories.query.filter_by(is_visible=True).one_or_none()
```
Comment (zgary) [inline app/feature/utils.py]: Remove if not being used.

File: app/feature/views.py [hunk 8]
```diff
@@ -10,9 +10,10 @@
 from app.feature.forms import FeaturedStoryForm, ModifyFeatureForm
 from app.feature.utils import create_featured_story, update_featured_story, hide_current_featured_story
 from app.models import Events, FeaturedStories, Stories
+from operator import attrgetter
 
 
-@feature.route('/', methods=['GET'])
+@feature.route('/', methods=['GET','POST'])
```
Comment (zgary) [inline app/feature/views.py]: Remove `POST` if not being used.

File: app/feature/views.py [hunk 9]
```diff
@@ -21,9 +22,13 @@ def listing():
 
     :return: renders the 'feature.html' template with featured story list
     """
-    featured_stories = FeaturedStories.query.filter_by().all()
+    featured_stories = sorted(FeaturedStories.query.filter_by(is_visible=True).all(), key=attrgetter('rank'))
+    hidden_stories = FeaturedStories.query.filter_by(is_visible=False).all()
 
-    return render_template('feature/feature.html', featured_stories=featured_stories)
+    # if request.method == 'POST':
```
Comment (zgary) [inline app/feature/views.py]: Remove if not being used.

File: app/static/js/feature/modify.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    var requiredFields = ["title", "description"];
+    for (var i = 0; i < requiredFields.length; i++) {
+        $("#" + requiredFields[i]).attr("data-parsley-required","");
+    }
+
+    // Specify length requirement of certain fields
+    titleInput.attr("data-parsley-maxlength", 90);
+    descriptionInput.attr("data-parsley-maxlength", 395);
+
+    // var errorList = [];
```
Comment (zgary) [inline app/static/js/feature/modify.js]: Remove if not being used.

File: app/templates/main/home.html [hunk 11]
```diff
@@ -48,22 +104,22 @@ <h1>Help us elevate and fuel New York City&#8217;s women activists and inspire a
...
-                    </div>
-                    <div id="bws-text" class="col-md-7">
-                        <h2 id="bws-h1">{{ featured_story.story.activist_first }} {{ featured_story.story.activist_last }}</h2>
-                        <p id="bws-quote">"{{ featured_story.quote }}"</p>
-                    </div>
-                </div>
-            </div>
-        </a>
-    {% endif %}
+{#    {% if featured_story %}#}
```
Comment (zgary) [inline app/templates/main/home.html]: With the implementation of the carousel, this should be removed.

**Review by zgary — CHANGES_REQUESTED** [review]

File: app/feature/utils.py (hunk 3 above)
Comment (TraceyKong) [reply app/feature/utils.py]: When a new featured story is added, it appears first on the carousel by default. The admin can  also decide its rank so the ranks in the database need to be updated as well.

Comment (zgary) [reply app/feature/utils.py]: 👍

**Review by zgary — APPROVED** [review]
